### PR TITLE
Python pure model persistence

### PR DESCRIPTION
**Pull request by amn41**

### CODE REVIEW

**Comment by davisking** [comment]
Cool, thanks for the PR.   I assume you noticed there are conflicts with the main MITIE codebase?

**Comment by amn41** [comment]
Hey @davisking rebased off master so conflicts gone

**Comment by davisking** [comment]
Cool, thanks for the PR.  I'll review this in a few days.

File: mitielib/include/mitie.h [hunk 1]
```diff
@@ -396,6 +413,22 @@ extern "C"
...
+        const char* filename,
+        const char* fe_filename
+    );
+    /*!
+        requires
+            - filename == a valid pointer to a NULL terminated C string
+            - fe_filename == a valid pointer to a NULL terminated C string
+        ensures
+            - Reads a saved MITIE text categorizer object from disk and returns a
+              pointer to the text categorizer. filename must point to a serialised pure_model
```
Comment (davisking) [inline mitielib/include/mitie.h]: What is a pure_model?  The references to classes should use the full class name so it's clear what class is being referenced.

File: mitielib/include/mitie.h [hunk 2]
```diff
@@ -132,6 +132,23 @@ extern "C"
...
+        const char* fe_filename
+    );
+    /*!
+        requires
+            - filename == a valid pointer to a NULL terminated C string
+            - fe_filename == a valid pointer to a NULL terminated C string
+        ensures
+            - Reads a saved MITIE named entity extractor from disk and returns a pointer to
+              the entity extractor object.
+            - filname must point to a serialized pure_model object,
```
Comment (davisking) [inline mitielib/include/mitie.h]: What is a pure_model?  The references to classes should use the full class name so it's clear what class is being referenced.

File: mitielib/include/mitie.h [hunk 3]
```diff
@@ -446,6 +479,23 @@ extern "C"
...
+    /*!
+        requires
+            - filename == a valid pointer to a NULL terminated C string
+            - ner != NULL
+        ensures
+            - Saves part of the ner object to disk in a file with the given filename.  
+              Specifically, it saves everything except the feature extractor. 
+              Once this function finishes you will be able to read the ner object 
+              from disk by calling 
+              mitie_load_named_entity_extractor(filename,feature_extractor_filename).
```
Comment (davisking) [inline mitielib/include/mitie.h]: Shouldn't this refer to the pure version?   Also, it's probably a good idea if the pure load functions reference these functions so it's clear what it is the load functions are capable of loading.

File: mitielib/include/mitie.h [hunk 4]
```diff
@@ -478,6 +528,24 @@ extern "C"
...
+    );
+    /*!
+        requires
+            - filename == a valid pointer to a NULL terminated C string
+            - tcat != NULL
+        ensures
+            - Saves the given text categorizer pure model to disk (omitting the feature extractor)
+              in a file with the given filename.  Once this function
+              finishes you will be able to read the text categorizer object from disk by calling
+              mitie_load_text_categorizer(filename,fe_filename).
```
Comment (davisking) [inline mitielib/include/mitie.h]: Same here

File: mitielib/mitie.py [hunk 5]
```diff
@@ -235,13 +244,21 @@ def get_possible_ner_tags(self):
         num = _f.mitie_get_num_possible_ner_tags(self.__obj)
         return [to_default_str_type(_f.mitie_get_named_entity_tagstr(self.__obj, i)) for i in xrange(num)]
 
-    def save_to_disk(self, filename):
+    def save_to_disk(self, filename, pure_model=False):
         """Save this object to disk.  You recall it from disk with the following Python
         code: 
-            ner = named_entity_extractor(filename)"""
+            ner = named_entity_extractor(filename)
+        if object was saved with pure_model=True, then you must also pass a feature extractor
```
Comment (davisking) [inline mitielib/mitie.py]: This is a little confusing

File: mitielib/mitie.py [hunk 6]
```diff
@@ -649,26 +672,38 @@ def train(self):
...
+        if (isinstance(filename, ctypes.c_void_p)):
             self.__obj = filename
         else:
-            self.__obj = _f.mitie_load_text_categorizer(filename)
-        if self.__obj is None:
-            raise Exception("Unable to load text_categorizer detector from " + filename)
+            if (fe_filename is None):
+                self.__obj = _f.mitie_load_text_categorizer(filename)
+            else:
+                print(filename,fe_filename)
```
Comment (davisking) [inline mitielib/mitie.py]: Remove print call

File: mitielib/mitie.py [hunk 7]
```diff
@@ -649,26 +672,38 @@ def train(self):
...
 
     def __del__(self):
         self.__mitie_free(self.__obj)
 
-    def save_to_disk(self, filename):
+    def save_to_disk(self, filename,pure_model=False):
         """Save this object to disk.  You recall it from disk with the following Python
         code: 
-            tcat = text_categorizer(filename)"""
+            tcat = text_categorizer(filename)
```
Comment (davisking) [inline mitielib/mitie.py]: This is also a little confusing.  The way it's worded makes it sound as if 'you must pass a feature extractor' when saving in pure mode.

**Comment by amn41** [comment]
thanks for feedback @davisking - updated

**Comment by davisking** [comment]
Cool, thanks for another PR :)

**Pull request merged**